pkg/realizer/client/client.go

```diff
@@ -18,25 +18,34 @@ import (
 	"fmt"
 
 	corev1 "k8s.io/api/core/v1"
+	"k8s.io/client-go/discovery"
 	"k8s.io/client-go/rest"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 )
 
-type ClientBuilder func(secret *corev1.Secret) (client.Client, error)
+type ClientBuilder func(secret *corev1.Secret, needDiscovery bool) (client.Client, discovery.DiscoveryInterface, error)
 
 func NewClientBuilder(restConfig *rest.Config) ClientBuilder {
-	return func(secret *corev1.Secret) (client.Client, error) {
+	return func(secret *corev1.Secret, needDiscovery bool) (client.Client, discovery.DiscoveryInterface, error) {
 		config, err := AddBearerToken(secret, restConfig)
 		if err != nil {
-			return nil, fmt.Errorf("adding bearer token: %w", err)
+			return nil, nil, fmt.Errorf("adding bearer token: %w", err)
 		}
 
 		cl, err := client.New(config, client.Options{})
 		if err != nil {
-			return nil, fmt.Errorf("creating client: %w", err)
+			return nil, nil, fmt.Errorf("creating client: %w", err)
 		}
 
-		return cl, nil
+		var discoveryClient discovery.DiscoveryInterface
+		if needDiscovery {
+			discoveryClient, err = discovery.NewDiscoveryClientForConfig(restConfig)
```

Comment:
related to https://github.com/vmware-tanzu/cartographer/pull/660/files#r813367165, have you considered [`NewMemCacheClient`](https://github.com/kubernetes/client-go/blob/41e0447e772813a0942f06f64f9eaf07a70cbdf8/discovery/cached/memory/memcache.go#L223-L233)?

Reply:
I had not seen this- thanks! switched to using this